Mathlib/AlgebraicGeometry/Torus.lean

```diff
@@ -0,0 +1,418 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.Morphisms.Affine
+
+/-!
+# Algebraic torus
+
+## Main definitions
+
+- `AlgebraicGeometry.Torus`: `𝕋(n; S)` is the split algebraic torus of dimension `n` over `S`.
+- `AlgebraicGeometry.Torus.coord`: The standard coordinate functions on the torus.
+- `AlgebraicGeometry.Torus.homOfVector`:
+  The morphism `X ⟶ 𝕋(n; S)` given by a `X ⟶ S` and a choice of `n`-coordinate functions.
+- `AlgebraicGeometry.Torus.homOverEquiv`:
+  `S`-morphisms into `Spec 𝕋(n; S)` are equivalent to the choice of `n` invertible global sections.
+- `AlgebraicGeometry.Torus.SpecIso`: `𝕋(n; Spec R) ≅ Spec R[X̲, X̲⁻¹]`.
+
+-/
+
+open CategoryTheory Limits MvPolynomial
+
+noncomputable section
+
+namespace AlgebraicGeometry
+
+universe v u
+
+variable (n : Type v) (S : Scheme.{max u v})
+
+section Affine
+
+variable (R : Type*) (σ : Type*) [CommRing R] {S : Type*} [CommRing S]
+
+/-- `R[σ, σ⁻¹]`. This is the ring representing `S ↦ Fun(σ, Sˣ)`. -/
+abbrev TorusRing : Type _ :=
+  Localization (M := MvPolynomial σ R) (.closure (Set.range X))
```

Comment:
More generally, it could be interesting to formalize diagonalizable group schemes https://ncatlab.org/nlab/show/diagonalizable+group+scheme
If `H` is an abelian group, this is the group scheme which represents the functor which sends a `R`-algebra `A` to the morphisms from `Multiplicative H` to the units in `A`, and it is the spectrum of a free `R`-algebra which is a free `R`-module on `H`. Tori are the particular case `H` is free. There would necessarily be some intersection in the API of both tori and diagonalizable groups, but this degree of generality would already facilitate the study of endo/automorphisms of tori.
(Also, all of this could be moved to a new folder `AlgebraicGeometry.GroupScheme`.)

Reply:
Ah I somehow thought we need a definition of $$\mathbb{G}_m$$ before we could talk about diagonalizable group schemes but you are right. I'll try to generalize things.

Reply:
I think I'll retry after #21427 and (a group object version of) #21394 lands

Reply:
Both landed :wink: